### PR TITLE
Use LU decomposition in solve context

### PR DESCRIPTION
**Pull request by joschmitt**
Use (FF)LU decomposition in the solve context for `QQMatrix` and the finite field types and do as much as possible in flint.
I hope I did not mess up the pointers too badly.

### CODE REVIEW

File: src/flint/fmpq_mat.jl [hunk 1]
```diff
@@ -692,6 +693,205 @@ function Solve._can_solve_internal_no_check(A::QQMatrix, b::QQMatrix, task::Symb
...
+end
+
+function Solve._can_solve_internal_no_check_right(C::Solve.SolveCtx{QQFieldElem}, b::QQMatrix, task::Symbol)
+   bint = zero_matrix(FlintZZ, nrows(b), ncols(b))
+   db = FlintZZ()
+   ccall((:fmpq_mat_get_fmpz_mat_matwise, libflint), Nothing,
+        (Ref{ZZMatrix}, Ref{ZZRingElem}, Ref{QQMatrix}), bint, db, b)
+   yint = zero_matrix(FlintZZ, ncols(C), ncols(b))
+   p = inv(Solve.lu_permutation(C)).d .- 1
+   flag = ccall((:fmpz_mat_solve_fflu_precomp, libflint), Int,
```
Comment (thofma) [inline src/flint/fmpq_mat.jl:826]:
```suggestion
   flag = ccall((:fmpz_mat_solve_fflu_precomp, libflint), Cint,
```

Comment (thofma) [reply src/flint/fmpq_mat.jl:826]: Any `int` on the C-side must be a `Cint` here. There are more, in particular the `unit` business.

File: src/flint/gfp_fmpz_mat.jl [hunk 2]
```diff
@@ -398,3 +416,77 @@ function Solve._can_solve_internal_no_check(A::FpMatrix, b::FpMatrix, task::Symb
...
+################################################################################
+
+function lu!(P::Generic.Perm, x::FpMatrix)
+   P.d .-= 1
+
+   rank = ccall((:fmpz_mod_mat_lu, libflint), Int,
+                (Ptr{Int}, Ref{FpMatrix}, Int),
+                P.d, x, 0)
+
+  P.d .+= 1
```
Comment (thofma) [inline src/flint/gfp_fmpz_mat.jl:449]: indentation is a bit off

**Comment by thofma** [comment]
Looks very good, in particular the comments and diagrams for the LU decomposition!

There are only a few non-mathematical things to clean up.

File: src/flint/fmpq_mat.jl (hunk 1 above)
Comment (joschmitt) [reply src/flint/fmpq_mat.jl]: I turned some `Int` into `Cint` now, but I wasn't sure about some places. Maybe have a look again.

**Comment by joschmitt** [comment]
This is good to go now from my point of view.

**Review by thofma — APPROVED** [review]

**Pull request merged**